### PR TITLE
middleware: Add gRPC instrumentation utilities from dskit

### PR DESCRIPTION
**Pull request by aknuds1**
Copy gRPC instrumentation utilities from [dskit](https://github.com/grafana/dskit/tree/main/middleware), as suggested offline by @bboreham.

### CODE REVIEW

File: middleware/grpc.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+	"github.com/prometheus/client_golang/prometheus"
+	grpcUtils "github.com/weaveworks/common/grpc"
+	"github.com/weaveworks/common/httpgrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// PrometheusGRPCUnaryInstrumentation records duration of gRPC requests client side.
+func PrometheusGRPCUnaryInstrumentation(metric *prometheus.HistogramVec) grpc.UnaryClientInterceptor {
```
Comment (bboreham) [inline middleware/grpc.go:17]: The equivalent server call in this package is `UnaryServerInstrumentInterceptor`, so perhaps `UnaryClientInstrumentInterceptor`?

File: middleware/grpc.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	return func(ctx context.Context, method string, req, resp interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		start := time.Now()
+		err := invoker(ctx, method, req, resp, cc, opts...)
+		metric.WithLabelValues(method, errorCode(err)).Observe(time.Since(start).Seconds())
+		return err
+	}
+}
+
+// PrometheusGRPCStreamInstrumentation records duration of streaming gRPC requests client side.
+func PrometheusGRPCStreamInstrumentation(metric *prometheus.HistogramVec) grpc.StreamClientInterceptor {
```
Comment (bboreham) [inline middleware/grpc.go:27]: The equivalent server call in this package is `StreamServerInstrumentInterceptor`, so perhaps `StreamClientInstrumentInterceptor`?

**Review by bboreham — COMMENTED** [review]
Bit more context: that code was moved from https://github.com/cortexproject/cortex/blob/ff858460bbb36cbe4708183250f7448242f8d928/pkg/util/middleware/grpc.go
about three months ago, then extended by https://github.com/grafana/dskit/pull/68.

I suggested putting the code here as it is the client equivalent of functions in https://github.com/weaveworks/common/blob/master/middleware/grpc_instrumentation.go.

Given the above, `grpc.go` is not a very explanatory name for the new files.  Maybe `grpc_client_instrumentation.go` ?  Or maybe just put the code in `grpc_instrumentation.go` ?

File: middleware/grpc.go (hunk 1 above)
Comment (aknuds1) [reply middleware/grpc.go]: @bboreham Thanks, done.

File: middleware/grpc.go (hunk 2 above)
Comment (aknuds1) [reply middleware/grpc.go]: @bboreham Thanks, done.

**Comment by aknuds1** [comment]
@bboreham Thanks for the review, I renamed the functions and moved them into middleware/grpc_instrumentation.go .

File: middleware/grpc_instrumentation.go [hunk 3]
```diff
@@ -45,3 +47,93 @@ func StreamServerInstrumentInterceptor(hist *prometheus.HistogramVec) grpc.Strea
...
+	if errResp, ok := httpgrpc.HTTPResponseFromError(err); ok {
+		statusFamily := int(errResp.Code / 100)
+		respStatus = strconv.Itoa(statusFamily) + "xx"
+	} else if grpcUtils.IsCanceled(err) {
+		respStatus = "cancel"
+	} else {
+		respStatus = "error"
+	}
+
+	return respStatus
```
Comment (pracucci) [inline middleware/grpc_instrumentation.go:138]: [nit] I think we can take the chance to further simplify this, removing `respStatus` at all:

```suggestion
	if err == nil {
		return "2xx"
	}

	if errResp, ok := httpgrpc.HTTPResponseFromError(err); ok {
		statusFamily := int(errResp.Code / 100)
		return strconv.Itoa(statusFamily) + "xx"
	} else if grpcUtils.IsCanceled(err) {
		return "cancel"
	} else {
		return "error"
	}
```

**Review by pracucci — APPROVED** [review]
FWIW, LGTM :)

File: middleware/grpc_instrumentation.go (hunk 3 above)
Comment (aknuds1) [reply middleware/grpc_instrumentation.go]: Applied, thanks.

**Review by bboreham — APPROVED** [review]
Sorry, fell off my radar.  LGTM.

**Pull request merged**